### ISSUE TITLE
Textmate grammar does not handle more than one location on a single line

### ISSUE THREAD

**Issue by rodinaarssen** [usethesource/rascal-language-servers#432]
**Describe the bug**

The Textmate grammar breaks Rascal's syntax highlighting if more than one location literal with three consecutive slashes appear "consecutively" (i.e., without a location literal with just two consecutive slashes in-between).

**To Reproduce**

![image](https://github.com/user-attachments/assets/f94c9e96-99a0-4747-8ac6-43a41ac982c9)
Note that from the second slash in the second literal

**Desktop (please complete the following information):**
 - Context: VS Code extension
 - Rascal Version 0.40.8-SNAPSHOT (~current main)
 - Rascal-lsp version: 0.12.0-head (current main)

**Comment by sungshik** [comment]
Thank you for reporting. Here's what happens:

  - For instance, on this line, the TextMate tokenizer assigns the following scopes:
    
    ```rascal
    if (|unknown:///| == l && |unknown:///| == l) {
    //             ^^^^^^^^^^^^^^^^^^^^^            constant.regexp
    //                                  ^^^^^^^^^^^ comment
    ```

  - The semantic highlighter subsequently tries to overwrite these scopes to `meta.embedded` to remove the formatting (using our own `uncategorized` semantic token type):
    
    ![image](https://github.com/user-attachments/assets/15b893bb-6c09-4969-af18-a13fcc1f6284)

  - **Issue 1:** However, while the overwrite does remove the color (`comment` is a different shade of gray in this theme), it doesn't remove italic, despite the ~~`fontStyle: italic`~~ in the token/scope inspector.

  - **Issue 2:** Moreover, VS Code seems to use only scopes assigned by the TextMate tokenizer to determine which brackets are included/excluded to match brackets. Brackets in a `comment` scope are excluded. As a result, the closing bracket on line 116 in the screenshot is not properly matched to the opening bracket. This is the same "related issue 2" in [this comment](https://github.com/usethesource/rascal-language-servers/issues/431#issuecomment-2324162495).

**Comment by DavyLandman** [comment]
It looks like we should be opening an issue on microsoft/vscode repo about this interaction bteween the textmate & the semantic tokens. Or not @sungshik ?

**Comment by sungshik** [comment]
I agree. I think the matching+highlighting of brackets in comments in the current release of the extension (i.e., before TextMate) is a symptom of the same issue. E.g., the purple brackets of `(c)` on line 2 in this screenshot:

![image](https://github.com/user-attachments/assets/ea2ad5ea-8daa-4766-9770-f0f554eda078)

Do you know if anyone looked into this before? #214 seems related (but not exactly the same).

**Comment by DavyLandman** [comment]
I've briefly looked into the brackets to see that there didn't seem to be a nice way to control where it shouldn't be active.

**Comment by sungshik** [comment]
Ok, [issue](https://github.com/microsoft/vscode/issues/229530) submitted.

**Comment by RedCMD** [comment]
and not just update the TextMate grammar?

**Comment by DavyLandman** [comment]
Also, we can have interpolation inside of strings:

```
loc validRascal(int i, str s) {
     return |file:///x/a/< "dir< i + 1 >" >/file< i > 2 ? s : "fallback" >.txt|; 
}
```

**Comment by sungshik** [comment]
Thanks, @RedCMD, for these questions and comments! Let me provide a bit more background about the work we are doing in relation to this particular issue.

The general context is Rascal, a meta-programming language/tools to develop (domain-specific) languages. First, the Rascal language offers high-level primitives to manually define grammars, code analyses, code transformations, etc., using a single notation. Next, the Rascal tools automatically generate parsers, type checkers, code generators, etc., from these definitions, possibly embedded in an LSP server + VS Code extension.

In previous versions of LSP servers + VS Code extensions generated using Rascal, the only form of highlighting offered was semantic highlighting: the LSP server would run the generated parser each time a document update was received and generate semantic tokens accordingly. There was no TextMate support in the VS Code extension.

To improve this, last year, we have been working on adding such TextMate support. Instead of asking Rascal users to manually define TM grammars, the idea is to automatically convert Rascal grammars to TM grammars. Such auto-conversions align well with Rascal's design principles. However, developing the conversion algorithm has been challenging, and the generated TM grammars are sometimes more complicated than hand-written TM grammars. Moreover, as Rascal grammars are context-free, there is little hope in general that the generated TM grammars will be 100% accurate.

For our use case, though, this is a fair compromise:
  - Rascal users still need to write only a Rascal grammar, without having to know anything about TextMate.
  - If the TextMate tokenizer misclassifies a token because the generated TM grammar is inaccurate, then there is always the semantic tokenizer to correct the mistake.

The issue here reported by @rodinaarssen is that:
 1. The TM grammar generated for the Rascal grammar *of the Rascal language itself* has inaccuracies, so the TextMate engine makes a tokenization error.
 2. The semantic tokenizer computes a "compensatory" semantic token, but it isn't processed by VS Code in the way we thought it would.

To fix the tokenization error, we could indeed manually change the TM grammar, but it would address only the symptom instead of the cause (in the grander scheme of things, given the broader context of what we are trying to achieve). 

I hope this clarifies the issue. Let me know if you have any further questions, and I would be happy to answer!

**Comment by RedCMD** [comment]
I've created a new TextMate Grammar 🎉
handles all the files in the rascal repository 
performance is good

still got a few tweaks to fix
imprpve performance for concrete

still a little bit of ambiguity between certain featues:
* tuples `<>`
* colons `:` especially around switch cases
* Types
* function declaration vs calls vs tree vs types

where should I send the PR?
* VSCode extension 
* Github linguist
* Shikijs package

**Comment by RedCMD** [comment]
here's my new extension https://marketplace.visualstudio.com/items?itemName=RedCMD.rascal-syntax-highlighter
